### PR TITLE
Integrate contracts v2.2.0

### PR DESCRIPTION
**Pull request by mariacarmina**
Fixes # .

Changes proposed in this PR:

- integrated factory contract
- updated contract version
- created test script for sapphire

### CODE REVIEW

File: src/@types/NFTFactory.ts
Comment (alexcos20) [inline src/@types/NFTFactory.ts:28]: why do we need this?

File: src/contracts/AccessList.ts
Comment (alexcos20) [inline src/contracts/AccessList.ts:70]: mint new token, or add address to accesslist

it's not minting another erc721 contract

File: src/contracts/AccessList.ts
Comment (alexcos20) [inline src/contracts/AccessList.ts:128]: delete address from accessList    or  burn tokenId

File: src/contracts/AccessListFactory.ts
Comment (alexcos20) [inline src/contracts/AccessListFactory.ts:42]: I would advice you to delete that typing (AccessListData) and use function params

File: src/contracts/NFT.ts
Comment (alexcos20) [inline src/contracts/NFT.ts:78]: only templateID == 4 supports accessLists , so pls make sure that we don't pass it to other template, because those templates might use those fields for something else

File: test/integration/PublishEditConsume.test.ts
Comment (alexcos20) [inline test/integration/PublishEditConsume.test.ts:352]: Out of curiosity, why did we removed this tests?

File: test/config.ts
Comment (alexcos20) [inline test/config.ts:52]: Maybe it's a good ideea to have two functions:
 - getAddressesForChain(chainID)
 - getAddressesForSapphire which calls getAddressesForChain

File: test/integration/PublishEditConsume.test.ts
Comment (mariacarmina) [reply test/integration/PublishEditConsume.test.ts:352]: I have commented them for debug because these were failing.

File: src/contracts/AccessListFactory.ts
Comment (mariacarmina) [reply src/contracts/AccessListFactory.ts:42]: ok sure.

File: src/contracts/AccessList.ts
Comment (mariacarmina) [reply src/contracts/AccessList.ts:70]: I'll modify

File: src/@types/NFTFactory.ts
Comment (mariacarmina) [reply src/@types/NFTFactory.ts:28]: it's a typing as `NftCreateData` is used when creating a NFT.

File: test/config.ts
Comment (mariacarmina) [reply test/config.ts:52]: I have that locally, not pushed yet.

File: test/integration/PublishEditConsume.test.ts
Comment (alexcos20) [reply test/integration/PublishEditConsume.test.ts:352]: well, you commented the order , but left the downloads on next tests :) and those are failing without orders..

Comment (mariacarmina) [reply test/integration/PublishEditConsume.test.ts:352]: I can focus on this problem, now since I am blocked on testing the integration side.

**Comment by alexcos20** [comment]
why for every call to accesslist functions do i need to provide accessListAddress ?

just have it in the constructor and do something like:
```
accessList = new AccessListContract('0x123')
owner = await accessList.getOwner()
```

**Comment by mariacarmina** [comment]
> why for every call to accesslist functions do i need to provide accessListAddress ?
> 
> just have it in the constructor and do something like:
> 
> ```
> accessList = new AccessListContract('0x123')
> owner = await accessList.getOwner()
> ```

From my point of view to pass the address for every call since we have methods related to allow list and deny list, because right now they are all together. If you want, we can have 2 children classes, one for allow list and one for deny list which can store the address in the constructor.

**Comment by mariacarmina** [comment]
> why for every call to accesslist functions do i need to provide accessListAddress ?
> 
> just have it in the constructor and do something like:
> 
> ```
> accessList = new AccessListContract('0x123')
> owner = await accessList.getOwner()
> ```

It wasn't a specific preference, I can change that AccessList class to extend SmartContractWithAddress class and store the address in constructor. The only reason for using this approach is that Nft class has in each method nft address passed as parameter.

**Comment by jamiehewitt15** [comment]
All the integration tests passed on the most recent run. The error message reports a naming conflict in the upload of `coverage` - perhaps try just changing the name to `integration-coverage`?

**Comment by jamiehewitt15** [comment]
The issue with the tests, is likely because `upload-artifact@v2` is deprecated. Can you update it to `upload-artifact@v4` in `.github/workflows/ci.yml`

File: .github/workflows/ci.yml
Comment (jamiehewitt15) [inline .github/workflows/ci.yml:148]: We're uploading three artifacts all with the same name and you're getting a name conflict error. 

I suggest changing this one:

```suggestion
          name: coverage-integration
```



And change the one for the unit test to `coverage-unit`

**Comment by jamiehewitt15** [comment]
You're getting `Error: Timeout of 40000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.` in the integration tests. Have you already tried rerunning the tests to see if it's a one time issue?

Also, you can try increasing the timeout to see if this fixes the issue.

**Comment by jamiehewitt15** [comment]
Everything is passing now, I just restarted the tests and it works.

**Comment by mariacarmina** [comment]
> You're getting `Error: Timeout of 40000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.` in the integration tests. Have you already tried rerunning the tests to see if it's a one time issue?
> 
> Also, you can try increasing the timeout to see if this fixes the issue.

tried that previously, increased to 60000 ms, it was related with coverage reports.

**Comment by mariacarmina** [comment]
Screenshot 2024-09-15 at 17 32 08

**Comment by mariacarmina** [comment]
set file object with encrypted format (the last test): https://explorer.oasis.io/testnet/sapphire/tx/0x2b8ce371f987a3e0309555973b677a75a3307fbe13855c4e43ae23bc666311ce

**Comment by mariacarmina** [comment]
The integration tests for ordering different types of assets is an older issue which pops up due to exceeded timeout. I commented the assets which rely on external services such as ipfs, arweave and create an issue for fixing those tests in the future since they are not impacted by this PR changes.

**Review by jamiehewitt15 — COMMENTED** [review]
Why are the sapphire integration tests in a folder called `scripts`? Seems like it would be better placed in `/test/integration/`

File: .gitignore
Comment (jamiehewitt15) [inline .gitignore:7]: Why have you done this? Before we ignored all javascript files in the test directory, now you want only to ignore js files if they are in either `/test/unit` or `test/integration`? I don't see any benefit to changing this

File: README.md
Comment (jamiehewitt15) [inline README.md:133]: This is worded poorly, I think what you are trying to say is: 

```suggestion
We are currently using the live Oasis Sapphire Test network for the integration tests. Please export the `PRIVATE_KEY` before running the tests.
```

File: test/scripts/SapphireIntegration.test.ts
Comment (jamiehewitt15) [inline test/scripts/SapphireIntegration.test.ts:18]: Given that these tests are using a private key on a live network, at some point, the wallet address is going to run out of tokens and the tests will fail. Is there some sort fo longer-term plan here, like getting the tests to work with barge? If so, do we have an issue for this?

File: package.json
Comment (bogdanfazakas) [inline package.json:33]: lets move the test file to its proper location /test/integration/
also renaming it to Sapphire.test.ts would be nice
then update the test:saphire to  "npm run mocha -- 'test/integration/Sapphire.test.ts'" .. and then just ignore the newloy moved file in the test:integration comand

File: src/contracts/AccessList.ts
Comment (bogdanfazakas) [inline src/contracts/AccessList.ts:90]: returns the transaction receipt

File: src/contracts/AccessList.ts
Comment (bogdanfazakas) [inline src/contracts/AccessList.ts:141]:
```suggestion
   * @return {Promise<ReceiptOrEstimate>} returns the transaction receipt or the estimateGas value
```

File: src/contracts/AccessList.ts
Comment (bogdanfazakas) [inline src/contracts/AccessList.ts:116]:
```suggestion
   * @return {Promise<ReceiptOrEstimate>} returns the transaction receipt or the estimateGas value
```

File: src/contracts/AccessList.ts
Comment (bogdanfazakas) [inline src/contracts/AccessList.ts:164]:
```suggestion
   * @return {Promise<ReceiptOrEstimate>}  returns the transaction receipt or the estimateGas value
```

File: src/contracts/AccessListFactory.ts
Comment (bogdanfazakas) [inline src/contracts/AccessListFactory.ts:78]: what if the tx fails, need to provide a proper response for the client otherwise the line bellow  events.args[0] will fail with uncaught error

File: src/utils/Asset.ts
Comment (bogdanfazakas) [inline src/utils/Asset.ts:19]: i would also add a default here if the address file is not present and remove the condition above 

```suggestion
    const data = JSON.parse(
    // eslint-disable-next-line security/detect-non-literal-fs-filename
    fs.readFileSync(
      process.env.ADDRESS_FILE ||
        `${homedir}/.ocean/ocean-contracts/artifacts/address.json`,
      'utf8'
    )
  )
```

File: src/config/ConfigHelper.ts
Comment (paulo-ocean) [inline src/config/ConfigHelper.ts:162]: this is misspelled again, chain name is: `oasis_sapphire_testnet`

File: src/utils/Asset.ts
Comment (paulo-ocean) [reply src/utils/Asset.ts:19]: can't see anything wrong here, what are you exactly suggesting @bogdanfazakas ?
this is perfectly fine
```
export function getOceanArtifactsAdresses(): any {
  try {
    if (process.env.ADDRESS_FILE) {
      // eslint-disable-next-line security/detect-non-literal-fs-filename
      const data = fs.readFileSync(process.env.ADDRESS_FILE, 'utf8')
      return JSON.parse(data)
    }
    return addrs
  } catch (error) {
    return addrs
  }
}
```
The `process.env.ADDRESS_FILE` should be the file path
if not defined or wrongly defined it returns the defaults from address.json, you're like repeating the same thing twice

File: .gitignore
Comment (mariacarmina) [reply .gitignore:7]: It was when I created a .js file, I'll correct it now since I have the .ts script, not .js

File: test/scripts/SapphireIntegration.test.ts
Comment (mariacarmina) [reply test/scripts/SapphireIntegration.test.ts:18]: That is the reason that I did not include the tests in the CI, because everytime the account will be drained, we'll have to top up the funds for those. For the moment it's not tested with CI, only locally. We can create an issue for it. @mihaisc suggested to have a separate job which not runs at every push but he said to leave this problem for now, I'll create an issue for it.

File: package.json
Comment (mariacarmina) [reply package.json:33]: Ok, sure I can do that.

File: src/contracts/AccessListFactory.ts
Comment (mariacarmina) [reply src/contracts/AccessListFactory.ts:78]: Yes, the same should be applied in NFTFactory when creating a NFT, try-catch block is missing from there too.

File: src/contracts/Datatoken4.ts
Comment (paulo-ocean) [inline src/contracts/Datatoken4.ts:180]: We need to refactor all these occurrences of `serviceID` (including comments).. It definitely is NOT the service id, but the service index.. so `serviceIndex` instead, otherwise we get easily confused

Comment (mariacarmina) [reply src/contracts/Datatoken4.ts:180]: right thing, sure, do that now.

File: src/contracts/AccessListFactory.ts
Comment (paulo-ocean) [inline src/contracts/AccessListFactory.ts:37]: lets kill this comment about AccessListData :-)

File: test/scripts/SapphireIntegration.test.ts
Comment (jamiehewitt15) [reply test/scripts/SapphireIntegration.test.ts]: ok cool, fair enough

File: src/utils/Asset.ts
Comment (bogdanfazakas) [reply src/utils/Asset.ts:19]: yep makes sense then, we can close this then

File: src/contracts/Datatoken4.ts
Comment (paulo-ocean) [inline src/contracts/Datatoken4.ts:35]: is the template only for confidential EVM sapphire?
Cause we're wrapping the signer with their sdk call `sapphire.wrap(...)` ..
and we have a few functions doing it again like:
`confidentialEVM === true ? sapphire.wrap(this.signer) : this.signer, ...`
So, we get a "double" wrap() for confidential EVM and a wrap() for any other EVM using this template 
For now, i would just remove the wrap() call on the constructor, and later, once we merge the other PR we can refactor this a bit more (from the config we would know if is confidential evm and if we should wrap the calls

Comment (mariacarmina) [reply src/contracts/Datatoken4.ts:35]: I can remove the wrapper from constructor to avoid confusions.

Comment (mariacarmina) [reply src/contracts/Datatoken4.ts:35]: Template 4 is only for confidential EVM, I mean, the scope of its creation is destinated for sapphire and files encryption on sapphire chain.

File: src/contracts/AccessListFactory.ts
Comment (mariacarmina) [reply src/contracts/AccessListFactory.ts]: created https://github.com/oceanprotocol/ocean.js/pull/1853

Comment (bogdanfazakas) [reply src/contracts/AccessListFactory.ts]: yes but the logic here still need few more changes, what if the tx fails, there would be no events matching the name we pass, so for at line 89 you will end up with an "cannot read property x of undefined"

Comment (mariacarmina) [reply src/contracts/AccessListFactory.ts]: I put a check only for `tx` returned by `sendTx`, because if that is not present the code below will fail.

Comment (bogdanfazakas) [reply src/contracts/AccessListFactory.ts]: i dont't really get it maybe i'm missing something, so if sendTx fails it will only log an error cause here we don't have a try catch or something handling that, so it go on wi calling `tx.wait()`, then trying to access events for that tx etc

**Review by jamiehewitt15 — APPROVED** [review]
Looks good to me

**Review by paulo-ocean — APPROVED** [review]
lgtm

**Comment by mariacarmina** [comment]
@alexcos20 can you check this PR please?

File: src/contracts/AccessListFactory.ts
Comment (mariacarmina) [reply src/contracts/AccessListFactory.ts]: if `sendTx` will fail, either will throw an error that will be caught in the try-catch block within `sendTx` function or the try-catch block from `deployAccessListContract` function which is surrounded by, either the tx receipt is null. If tx returned from `sendTx` does not exist, this will throw an error with message: `Tx for deploying new access list was not processed on chain.`. It will not go to `tx.wait()` please check my latest code where I have added the condition, because it is above `tx.wait()`.

**Review by alexcos20 — APPROVED** [review]

**Pull request merged**
